### PR TITLE
Backups filesets: specify what to include, exclude, how long to keep

### PR DESCRIPTION
**Pull request by weedySeaDragon**
### PT Story: backups: how long to keep files on production? split up data v. code
https://www.pivotaltracker.com/story/show/166749266

### Changes proposed in this pull request:
Can now specify a "file set" in the backup configuration.
    A "FileSet" has a `name`, a list of `files` to be backed up, and the number of `days_to_keep` this file.  Patterns for exclusion can be specified in `excludes`.

This is what is now in `load_conditions`:
```
{ class_name: 'Backup',
          timing:     :every_day,
          config:     { days_to_keep:     { code_backup: 4,
                                            db_backup:   15,
                                            files_backup: 31},
                        backup_directory: nil,
                        filesets: [
                            {name: 'logs',
                             days_to_keep: 8,
                             files: [RUNNING_LOG, NGINX_LOG_DIR, File.join(APP_DIR, 'log')]
                            },
                            {name: 'code',
                             days_to_keep: 3,
                             files: [APP_DIR],
                             excludes: ['public', 'docs', 'features', 'spec','tmp', '.yardoc']
                            },
                            {name: 'app-public',
                             days_to_keep: 3,
                             files: [PUBLIC_DIR]
                            },
                            {name: 'config env secrets',
                            days_to_keep: 32,
                             files: [File.join(APP_DIR, 'config', '*.yml'), '.env']
                            }
                        ]
          }
        },
```

#### Specifics:
1.  change FilesBackupMaker -> FileSetBackupMaker
2. updated the Backup configuration in `load_conditions`
3. Added an Acceptance Test in `backups_spec.rb` that specifically creates many directories and has FileSets in the configuration, including some that have `excludes.`  The test verifies that the backup `tar` files actually have the correct files and not those that should be excluded.

### Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

File: app/models/conditions_response/backup.rb [hunk 1]
```diff
@@ -200,7 +94,7 @@ def self.condition_response(condition, log)
 
   rescue Slack::Notifier::APIError => slack_error
     # Halt the backup if we cannot write to Slack; log then raise the error
-     log_slack_error(slack_error, log, '(in rescue at bottom of condition_response)')
+    log_slack_error(slack_error, log, '(in rescue at bottom of condition_response)')
```
Comment (patmbolger) [inline app/models/conditions_response/backup.rb]: I'm not sure there's much value in logging an error at this level for a slack error.  That is, why do we need to know where we were in the condition processing when the slack error occurred, when the exception itself had nothing to do with this condition's processing?  

The condition processing task will rescue the re-raised exception, log an error, and then retry the specific condition.

File: app/models/conditions_response/shf_backup_makers/abstract_backup_maker.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    # Set the backup target and the backup sources
+    def initialize(target_filename: base_filename,
+                   backup_sources: default_sources)
+      @target_filename = target_filename
+      @backup_sources = backup_sources
+    end
+
+
+    # Do the backup. Default target is the target_filename; default sources = the backup sources)
+    def backup(target: target_filename, sources: backup_sources)
```
[the remark this answers is not in the document]
Comment (patmbolger) [reply app/models/conditions_response/shf_backup_makers/abstract_backup_maker.rb:45]: Also, we're using an instance var - `@slack_error_already_logged` - to keep track of whether we've already seen and logged a slack exception.  

This means that now _each_ condition processing class must keep track of whether it. has seen, and logged, a slack exception.

This should the the responsibility of the processing task only.  With that, an argument could be added to the `condition_response` method that tells the condition response class whether or not to try to send notification to slack.

**Review by patmbolger — COMMENTED** [review]
Please see comments.

File: app/models/conditions_response/backup.rb (hunk 1 above)
Comment (weedySeaDragon) [reply app/models/conditions_response/backup.rb]: I see your point but still believe that if you have an error, you need to report exactly where that happened. We have more information at this exact moment (where the Slack error happened during this condition's processing), so we should provide it, not disappear it.

If an error happens, we need to provide information about exactly when and where it really happened.  Doing anything that obscures either of those pieces of information will cause pain during debugging. I agree that the error is about communicating with Slack, but we need to be precise about when the error occurred.  This provides a _timestamp_ when the error occurred  that gives us the information needed to check with what happened to communication with Slack.

File: app/models/conditions_response/shf_backup_makers/abstract_backup_maker.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (weedySeaDragon) [reply app/models/conditions_response/shf_backup_makers/abstract_backup_maker.rb:45]: Agreed.  I just hadn't made those changes, trying to keep the scope of this PR relatively focused.

That will require a change to the `condition_response` method signature, affecting all of the conditions, yes?

[the remark this answers is not in the document]
Comment (patmbolger) [reply app/models/conditions_response/shf_backup_makers/abstract_backup_maker.rb:45]: Yes, an optional argument added to that method should do it.

File: app/models/conditions_response/backup.rb (hunk 1 above)
Comment (patmbolger) [reply app/models/conditions_response/backup.rb]: We'll get the timestamp from the log entry entered by the rake task.  But in any case I don't feel too strongly about this.

File: spec/shared_context/expect_tar_has_entries.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+RSpec.shared_context 'expect tar file has entries' do
+
+# use 'tar --list' to verify that the tar_file has the expected files and directories
+  def expect_tar_has_these_entries(tar_file, expected_entries)
+
+    expect(File.exist?(tar_file)).to be_truthy
+
+    backup_file_list = %x<tar --list --file=#{tar_file}>
+    backup_file_list.gsub!(/\n/, ' ')
```
[the remark this answers is not in the document]
Comment (weedySeaDragon) [reply spec/shared_context/expect_tar_has_entries.rb:9]: tr! does not work with Regexps

File: app/models/conditions_response/shf_backup_makers/abstract_backup_maker.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (weedySeaDragon) [reply app/models/conditions_response/shf_backup_makers/abstract_backup_maker.rb:45]: I'll add a comment where it's used, but `@slack_error_already_logged` is used to ensure that a Slack error is not raised twice _only within Backup_.  This is different from whether or not to use Slack Notification at all.
If a slack error happens within a loop (e.g. while uploading to AWS S3), `@slack_error_already_logged`  ensures that the Slack error is not also logged when it is rescued at the end of the `condition_response` method.

File: lib/tasks/conditions/process_conditions.rake [hunk 4]
```diff
@@ -32,7 +32,7 @@ namespace :shf do
 
         if use_slack_notification
           SHFNotifySlack.notify_after(class_name) do
-            process_klass(klass, condition, log)
+            process_klass(klass, condition, log, use_slack_notification: use_slack_notification)
```
Comment (patmbolger) [inline lib/tasks/conditions/process_conditions.rake:35]: I could be wrong but it appears there might be a logic problem in the slack-notification error handling.

The rescue clause sets `use_slack_notification = false`, logs error(s) and then calls `retry`.

Subsequently line 38 is called:
```
process_klass(klass, condition, log)
```

The signature of that method:
```
def process_klass(klass, condition, log, use_slack_notification: true)
```
will result in calling the specific condition's `condition_response` method with `use_slack_notification` set too true.

**Review by patmbolger — COMMENTED** [review]
Please see comment.

File: lib/tasks/conditions/process_conditions.rake (hunk 4 above)
Comment (weedySeaDragon) [reply lib/tasks/conditions/process_conditions.rake:35]: Good catch.

**Review by patmbolger — APPROVED** [review]

**Pull request merged**